generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -1418,6 +1421,91 @@ public boolean preconditionHolds(Expr expr, BasicType basicType, boolean constCo
     }
   }
 
+  /**
+   * Identity transformation to insert a vector or matrix into a larger vector or matrix using a
+   * type constructor, then extract it back out again using another type constructor. When
+   * performed, transforms an expression e with type t -> identity(t(identity(m(identity(e), ..)))),
+   * where m is a type of equal or greater width than t.
+   * The rules for this smaller -> larger -> smaller transformation are as follows:
+   *     If the given type is a vector, the inner constructor type can be a vector of the same
+   *     element type with equal/greater width than the given type, or (provided that the given type
+   *     is a floating point genType) it can be a matrix with equal/greater column width than the
+   *     given type.
+   *     If the given type is a matrix, the inner constructor type can be a matrix with
+   *     equal/greater column/row width than the given type.
+   */
+  private class IdentityCompositeConstructorExpansion extends AbstractIdentityTransformation {
+    private IdentityCompositeConstructorExpansion() {
+      super(BasicType.allNumericTypes().stream().filter(
+          item -> !item.isScalar()).collect(Collectors.toList()), true);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert !type.isScalar();
+      // Our inner constructor type will be equal or larger in width than our given type. The goal
+      // of the next loop is to populate this list with all types that can fit our given type.
+      final List<BasicType> innerConstructorTypes = new ArrayList<BasicType>();
+      for (BasicType constructorType : BasicType.allBasicTypes().stream().filter(
```

Comment:
This loop is a mess of conditionals, but I couldn't think of a more elegant solution since the conditions are pretty intertwined with each other, and combining if statements with && causes typing issues.

Reply:
I wonder whether something like this might be cleaner:

if (type is not matrix) {
  for i in [length of vector + 1, 4] {
    add vector type with same element type and base type i
  }
  if element type is float {
    add all relevant matrix types
  }
} else {
  // loops to work out which larger matrices to add
}

Reply:
This is much better - I was trying to have one all-encompassing loop, but breaking down and having multiple loops makes the logic much simpler.